streamz/sources.py

```diff
@@ -138,29 +138,60 @@ class from_kafka(Source):
             group, each message will be passed to only one of them.
     poll_interval: number
         Seconds that elapse between polling Kafka for new messages
+    asynchronous: bool (True)
+        In order to start polling on a callback
+    loop: None or IOLoop
+        Loop in which to run polling as a coroutine
 
     Example
     -------
 
     >>> source = Stream.from_kafka(['mytopic'],
-    ...        dict(url='localhost:9092', group='streamz'))  # doctest: +SKIP
+    ...           {'bootstrap.servers': 'localhost:9092',
+    ...            'group.id': 'streamz'})  # doctest: +SKIP
     """
-    def __init__(self, topics, consumer_params, poll_interval=0.1):
-        import confluent_kafka as ck
-        IOLoop.current().add_callback(self.poll_kafka)
+    def __init__(self, topics, consumer_params, poll_interval=0.1,
+                 asynchronous=True, **kwargs):
         self.cpars = consumer_params
-        self.consumer = ck.Consumer(consumer_params)
-        self.consumer.subscribe(topics)
+        self.consumer = None
         self.topics = topics
-        self.sleep = poll_interval
+        self.poll_interval = poll_interval
 
-        super(from_kafka, self).__init__()
+        super(from_kafka, self).__init__(asynchronous=asynchronous, **kwargs)
+        self._connect_consumer()
 
     @gen.coroutine
     def poll_kafka(self):
         while True:
-            msg = self.consumer.poll(0)
-            if msg is None or msg.error():
-                yield gen.sleep(self.sleep)
+            f = open('/Users/mdurant/status', 'a')
+            if self.consumer is not None:
+                msg = self.consumer.poll(0)
+                if msg is None or msg.error():
+                    f.write('error')
+                    yield gen.sleep(self.poll_interval)
+                else:
+                    f.write('value')
+                    yield self.emit(msg.value())
             else:
-                yield self.emit(msg.value())
+                f.write('no consumer')
+                yield gen.sleep(self.poll_interval)
+
+    def _connect_consumer(self):
+        open('/Users/mdurant/status', 'a').write('_connect')
+        import confluent_kafka as ck
+        import atexit
+        if self.loop is not None:
+            # need some IOLoop to run on
+            self.loop.add_callback(self.poll_kafka)
+        self.consumer = ck.Consumer(self.cpars)
+        self.consumer.subscribe(self.topics)
+        atexit.register(self._close_consumer)  # release on server
```

Comment:
Close is now called on both `del` and shutdown (but only once), so should not have problems with server saturation.

Reply:
`weakref.finalize` seems to be the way the cool kids do it these days.  This is done a few times in the dask.distributed codebase if you're interested in grepping there.

Reply:
You might want the following here:

```python
import weakref
weakref.finalize(self, consumer.close)
```

If there is some way that the consumer may already have been collected, then I would use a weakref

```python
import weakref

def close(ref):
    consumer = ref()
    if consumer is not None:
        consumer.close()

weakref.finalize(self, close, weakref.ref(self.consumer))
```

Reply:
Have you seen error like `TypeError: cannot create weak reference to 'cimpl.Consumer' object` before?

Reply:
That's odd.  No I haven't.  I recommend raising that upstream.  Something with their C-Extension class must be off.

Reply:
https://github.com/confluentinc/confluent-kafka-python/issues/327